### PR TITLE
[APINotes] Remove VersionedInfoRole

### PR DESCRIPTION
**Pull request by jrose-apple**
When API notes were first added, they would defer to the header when something was explicitly specified---for example, if the API notes had one set of nullability annotations from before the header was audited. This behavior didn't make sense for versioned API notes, however, where annotations needed for Swift 3 compatibility were intended to override the unversioned information in both the API notes and the header itself.

Now that the versioning of API notes is well-established, simplify things by having the API notes always "win"; that is, an annotation in API notes will always override what's in the headers.

While here, remove the obsolete `FactoryAsInit` key from the API notes dictionary. This never had a good implementation with versioned API notes anyway.

Getting this into a release branch or 'stable' depends on apple/swift#6901 and apple/swift#6907.

### CODE REVIEW

File: lib/APINotes/APINotesWriter.cpp [hunk 1]
```diff
@@ -766,7 +766,7 @@ namespace {
     }
 
     void emitUnversionedInfo(raw_ostream &out, const ObjCMethodInfo &info) {
-      uint8_t payload = info.FactoryAsInit;
+      uint8_t payload = 0;
```
Comment (DougGregor) [inline lib/APINotes/APINotesWriter.cpp:769]: I see that you kept the API notes binary format the same. Thanks!

File: lib/Sema/SemaDecl.cpp [hunk 2]
```diff
@@ -13537,7 +13537,6 @@ Decl *Sema::ActOnTag(Scope *S, unsigned TagSpec, TagUseKind TUK,
 
   if (Attr)
     ProcessDeclAttributeList(S, New, Attr);
-  ProcessAPINotes(New);
```
Comment (DougGregor) [inline lib/Sema/SemaDecl.cpp:13540]: Why did this go away?

**Review by DougGregor — CHANGES_REQUESTED** [review]
Why is it correct to remove the ProcessAPINotes call from ActOnTag? If it is correct, we should separate it out.

File: lib/Sema/SemaDecl.cpp (hunk 2 above)
Comment (jrose-apple) [reply lib/Sema/SemaDecl.cpp:13540]: Whoops! That was supposed to be a separate patch (with a good explanation and a test case). Thanks for catching this!

**Comment by jrose-apple** [comment]
Fixed!

**Review by DougGregor — APPROVED** [review]
Thanks!

**Pull request merged**